lib/emerald/parser.rb

```diff
@@ -63,6 +76,28 @@ def parse_string(source)
       rest_of_source = drop(source, string_range.size)
       [string, rest_of_source]
     end
+  
+    def parse_list(source)
+      pattern = /([\s\S]+)/
+      list_range = pattern.match(source).to_s
+      complete_list = matching_brackets(list_range)
+      list = List.new()
+      [list, complete_list]
+    end
+
+    def matching_brackets(list_range)
+      counter = 0
+      matched_list = []
+      list_range.each_char do |i|
+        counter += 1 if i == '(' 
+        counter -= 1 if i == ')'
+        matched_list.push(i)
+        break if counter == 0 
+      end
+      rest_of_source = drop(list_range, matched_list.size)
+      list_contents = matched_list.join("").to_s[1...matched_list.size-1]
+      [list_contents, rest_of_source]
+    end
```

Comment:
Here's a (slightly simpler) Lisp parser I wrote in Rust. Might give you some ideas :)

https://github.com/lpil/learning-rust/blob/master/lisp-calc/src/parser.rs#L48